xarray/backends/common.py

```diff
@@ -195,6 +198,37 @@ def __init__(self, writer=None):
             writer = ArrayWriter()
         self.writer = writer
 
+    def encode(self, variables, attributes):
```

Comment:
This is definitely cleaner than what we had before, but I am reluctant to give the idea that this this is a new supported method for third-party DataStore classes. Maybe we can call this `_encode` for now, or add a warning about implementing it?

Eventually, I would like to remove all implementations from the DataStore base classes, and leave them as purely abstract. This will make it clearer to new backend implementers what they actually should/can implement.

So instead of implementing an `encode()` method, data store classes could have a list of default `encoders` (see xarray.coding) used when reading/writing data. But xarray.coding isn't quite ready for this yet...

Reply:
Clearly some refactoring will be needed to this once the overall backend refactoring moves forward. For now, however, this seems like a reasonable compromise.

Reply:
@shoyer - I'm a bit confused here. As you'll see in the Zarr backend, the `encode_variable` method is applying a list of encoders. Where in the `WritableDataStore` were you envisioning the application of the encoders?

Reply:
@shoyer - What would you say to merging this in its current state and leaving the encoders refactor to a separate PR? I'm happy to make more changes here but a) I'm not sure how to address your last comment, and b) I've already drifted a fair ways off track with this PR.

Reply:
I'm fine with that. We may want to change it more in the future but this is a clear improvement for now.